code/site/libraries/authentication/key.php

```diff
@@ -29,15 +29,90 @@ class ApiAuthenticationKey extends ApiAuthentication
 	 */
 	public function authenticate()
 	{
-		$app = JFactory::getApplication();
-		$query_token = $app->input->get('key', '', 'STRING');
+		$app          = JFactory::getApplication();
+		$query_token  = $app->input->get('key', '', 'STRING');
 		$header_token = $this->getBearerToken();
-		$key = $header_token ? $header_token : $query_token;
-
-		$token = $this->loadTokenByHash($key);
+		$key          = $header_token ? $header_token : $query_token;
+		$token        = $this->loadTokenByHash($key);
 
 		if (isset($token->state) && $token->state == 1)
 		{
+			// Get user for this key
+			$user         = JFactory::getUser($token->userid);
+			$isSuperAdmin = $user->authorise('core.admin');
+
+			// If this user is super admin user
+			if ($isSuperAdmin)
+			{
+				$userToImpersonate = self::getUserToImpersonate();
+
+				// If other is to be impersonated
+				if ($userToImpersonate)
+				{
+					$searchFor      = '';
+					$searchForValue = '';
+
+					if (preg_match('/email:(\S+)/', $userToImpersonate, $matches))
+					{
+						$searchFor      = 'email';
+						$searchForValue = $matches[1];
+					}
+					elseif (preg_match('/username:(\S+)/', $userToImpersonate, $matches))
+					{
+						$searchFor      = 'username';
+						$searchForValue = $matches[1];
+					}
+					elseif (is_numeric($userToImpersonate))
+					{
+						$userId = $userToImpersonate;
+					}
+					else
+					{
+						ApiError::raiseError("400", JText::_('COM_API_USER_NOT_FOUND'), 'APIValidationException');
+
+						return false;
+					}
+
+					// If username or emailid exists ?
+					if ($searchFor)
+					{
+						$db = JFactory::getDbo();
```

Comment:
JTable::load() allows loading a record based on any field not just primary key, so you can skip this query ?

Reply:
JTable for users assumes first param to be integer only

```    
/**
	 * Method to load a user, user groups, and any other necessary data
	 * from the database so that it can be bound to the user object.
	 *
	 * @param   integer  $userId  An optional user id.
	 * @param   boolean  $reset   False if row not found or on error
	 *                           (internal error state set in that case).
	 *
	 * @return  boolean  True on success, false on failure.
	 *
	 * @since   1.7.0
	 */
	public function load($userId = null, $reset = true)
	{
```

so DB query is needed @coolbung